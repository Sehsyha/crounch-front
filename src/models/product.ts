export interface ProductResponse {
  products: ProductApi[];
}

export interface ProductApi {
  product_name: string;
  categories: string;
  code: string;
}

export interface Product {
  barCode: string;
  name: string;
  categories: string[];
}
