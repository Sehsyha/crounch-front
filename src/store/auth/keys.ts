export enum AuthActions {
  LOGIN = 'LOGIN',
  SIGNUP = 'SIGNUP',
  LOGOUT = 'LOGOUT',
}

export enum AuthGetters {
  IS_AUTHENTICATED = 'IS_AUTHENTICATED',

}

export enum AuthMutations {
  LOGIN = 'LOGIN',
  LOGOUT = 'LOGOUT',
}
