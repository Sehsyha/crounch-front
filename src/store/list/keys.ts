export enum ListKeys {
  CREATE = 'CREATE',
  CREATED = 'CREATED',
  GETOWNERS = 'GETOWNERS',

  ADD = 'ADD',
  SET = 'SET',
  RESET = 'RESET',
};
