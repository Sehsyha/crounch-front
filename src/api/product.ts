import { doFetch } from './api';
import { Product } from '@/models/product';

export function createProduct(name: string): Promise<Product> {
  return doFetch({
    url: 'products',
    method: 'POST',
    data: { name }
  });
}

export function searchProduct(name: string): Promise<Array<Product>> {
  return doFetch({
    url: 'products/search',
    method: 'POST',
    data: { name }
  })
}
