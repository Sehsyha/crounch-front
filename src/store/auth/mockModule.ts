import { AuthKeys } from './keys';
import { Module } from 'vuex';
import { AuthState } from '.';
import { RootState } from '..';

const actions = {
  [AuthKeys.SIGNUP]: jest.fn(),
  [AuthKeys.LOGIN]: jest.fn(),
  [AuthKeys.LOGOUT]: jest.fn(),
};

const getters = {
  isAuthenticated: jest.fn(),
};

export function createAuthModuleMock(): Module<AuthState, RootState> {
  return {
    namespaced: true,
    actions,
    getters,
    state: {
      token: '',
      status: '',
    },
  };
}
