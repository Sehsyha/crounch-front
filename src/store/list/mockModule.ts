import { Module } from 'vuex';

import { ListKeys } from './keys';
import { ListState } from '.';
import { RootState } from '..';

const actions = {
  [ListKeys.CREATE]: jest.fn(),
  [ListKeys.CREATED]: jest.fn(),
  [ListKeys.GETOWNERS]: jest.fn(),
};

const getters = {
  lists: jest.fn(),
};

const mutations = {
  [ListKeys.RESET]: jest.fn(),
};

export function createListModuleMock(): Module<ListState, RootState> {
  return {
    namespaced: true,
    actions,
    getters,
    mutations,
    state: {
      lists: [],
    },
  };
}
