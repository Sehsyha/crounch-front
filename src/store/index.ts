import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import { auth } from './auth';
import { ListModule } from './ListModule';
import { product } from './product'

Vue.use(Vuex);

export interface RootState { }

const store: StoreOptions<RootState> = {
  state: {},
  modules: {
    auth,
    ListModule,
    product,
  },
  strict: true,
};

export default new Vuex.Store<RootState>(store);
