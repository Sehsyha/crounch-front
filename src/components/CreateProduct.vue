<template>
  <div id="create-product">
    <h2>Création de produit</h2>
    <div v-if="!isNameValid" id="name-error">Le nom doit être renseigné</div>
    <div v-if="result !== ''" class="result">{{ result }}</div>
    <form @submit.prevent="createProduct">
      <input type="text" placeholder="Nom" v-model="name" />
      <input type="submit" value="Créer" />
    </form>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import { Action } from 'vuex-class';

import { ProductActions } from '@/store/product/keys';
import { productNamespace } from '@/store/product';

@Component
export default class CreateProduct extends Vue {
  @Action(ProductActions.CREATE, productNamespace) doCreateProduct: any;
  name: string = '';
  result: string = '';

  createProduct() {
    if (!this.isNameValid) {
      return;
    }

    this.doCreateProduct({ name: this.name }).then(() => {
      this.result = 'Créé';
      this.name = '';
    });
  }

  get isNameValid(): boolean {
    return this.name.length > 0;
  }
}
</script>
