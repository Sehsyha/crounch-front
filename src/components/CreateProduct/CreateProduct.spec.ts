import { Wrapper } from '@vue/test-utils';
import { Module } from 'vuex';

import { shallowComponent } from '@/utils/test';
import CreateProduct from './CreateProduct.vue';
import { ProductActions } from '@/store/product/keys';
import { createProductModuleMock } from '@/store/product/mockModule';
import { ProductState } from '@/store/product';
import { RootState } from '@/store';

jest.mock('@/store/product');

describe('CreateProduct', () => {
  let wrapper: Wrapper<CreateProduct>;
  let product: Module<ProductState, RootState>;
  const name = 'shopping list';

  beforeEach(() => {
    product = createProductModuleMock();

    const modules = {
      product,
    };

    wrapper = shallowComponent(CreateProduct, { modules });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('Should render.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('State', () => {
    it('Should initialize product name state to empty string.', () => {
      expect(wrapper.vm.$data.name).toBe('');
    });

    it('Should initialize result state to empty string.', () => {
      expect(wrapper.vm.$data.result).toBe('');
    });
  });

  describe('Form validation', () => {
    it('Should render an error when name is not valid.', () => {
      expect(wrapper.find('#name-error').exists()).toBeTruthy();
    });

    it('Should not render an error when name is valid.', () => {
      wrapper.setData({ name });

      expect(wrapper.find('#name-error').exists()).toBeFalsy();
    });
  });

  it('Should not dispatch create product store action when form is not valid.', () => {
    wrapper.find('[type=submit]').trigger('click');

    expect(product.actions![ProductActions.CREATE] as jest.Mock).not.toHaveBeenCalled();
  });

  it('Should dispatch create product store action when form is valid and submitted.', done => {
    wrapper.setData({ name });

    wrapper.find('[type=submit]').trigger('click');

    setTimeout(() => {
      expect(product.actions![ProductActions.CREATE] as jest.Mock).toHaveBeenCalledWith(
        expect.anything(),
        { name },
      );
      done();
    });
  });

  it('Should display result message when request succeed.', done => {
    wrapper.setData({ name });

    wrapper.find('[type=submit]').trigger('click');

    setTimeout(() => {
      expect(wrapper.find('.result').text()).toEqual('Créée')
      done();
    });
  });

  it('Should reset name when request succeed.', done => {
    wrapper.setData({ name });

    wrapper.find('[type=submit]').trigger('click');

    setTimeout(() => {
      expect(wrapper.vm.$data.name).toBe('');
      done();
    });
  });
});
