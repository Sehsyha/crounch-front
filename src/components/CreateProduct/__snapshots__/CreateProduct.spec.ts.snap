// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateProduct Should render. 1`] = `
<div id="create-product">
  <div id="name-error">Le nom doit être renseigné</div>
  <!---->
  <form><input type="text" placeholder="Nom"> <input type="submit" value="Créer"></form>
</div>
`;
